### PR TITLE
Cache RawData for Commit, Tag, & Tree, fixes #6

### LINKED ISSUE #6: Cache the cid
go-ipfs and friends make no attempt to cache Node CIDs and will call `Cid()` repeatedly as-needed. Therefore, calling `Cid()` *should not* recompute the CID each time. Instead, it should either be cached on when first computed (preferably, when the node is first decoded).

Furthermore, the raw data should be kept around if possible so that calling `RawData()` is free. I don't expect this to tie up *that* much memory.

### PR DESCRIPTION
**Pull request by sameer**
A `rawdata []byte` member is added to each type. It is computed on the first call to `RawData()`.
Also noticed a few odd indentations and fixed them.

Let me know if there are any changes that need to be made, feedback is welcome.

### CODE REVIEW

File: commit.go [hunk 1]
```diff
@@ -145,7 +151,9 @@ func (c *Commit) RawData() []byte {
 		fmt.Fprintln(buf, line)
 	}
 	fmt.Fprintf(buf, "\n%s", c.Message)
-	return buf.Bytes()
+	c.rawdata = buf.Bytes()
```
Comment (magik6k) [inline commit.go]: Needs a lock over this whole function to be threadsafe

File: tree.go [hunk 2]
```diff
@@ -16,11 +16,12 @@ type Tree struct {
 	size    int
 	order   []string
 	cid     cid.Cid
+	rawdata []byte
```
Comment (magik6k) [inline tree.go]: Appears to bu unused / not implemented here

File: commit.go (hunk 1 above)
Comment (sameer) [reply commit.go]: Ok, I used a [Once](https://golang.org/pkg/sync/#Once) to fix this.

File: tree.go (hunk 2 above)
Comment (sameer) [reply tree.go]: Whoops, fixed.

**Comment by sameer** [comment]
Sure, should that be done as a `*testing.B` in `git_test.go`, or should I just try it out against the linux kernel or something with a before & after?

**Comment by magik6k** [comment]
Whatever works best for you and you think should show a difference is fine.

**Comment by sameer** [comment]
Ok, I had a chance to do more comprehensive testing. All the files in the repo are already added to my ipfs daemon, so they should return fast. The bench script just does `rm -rf .git/ipld/ && git push ipld:: master`.
Before: 
```
 Performance counter stats for './bench.sh' (100 runs):

            554.14 msec task-clock:u              #    0.623 CPUs utilized            ( +-  0.85% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
             2,680      page-faults:u             # 4840.519 M/sec                    ( +-  1.63% )
     1,121,182,658      cycles:u                  # 2025221.108 GHz                   ( +-  0.29% )
     1,280,351,427      instructions:u            #    1.14  insn per cycle           ( +-  0.13% )
       227,040,471      branches:u                # 410109049.439 M/sec               ( +-  0.14% )
         4,581,354      branch-misses:u           #    2.02% of all branches          ( +-  0.32% )

            0.8896 +- 0.0205 seconds time elapsed  ( +-  2.31% )
```
After:
```
 Performance counter stats for './bench-fast.sh' (100 runs):

            567.96 msec task-clock:u              #    0.646 CPUs utilized            ( +-  0.86% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
             2,678      page-faults:u             # 4719.887 M/sec                    ( +-  1.80% )
     1,131,191,140      cycles:u                  # 1993463.988 GHz                   ( +-  0.33% )
     1,281,825,576      instructions:u            #    1.13  insn per cycle           ( +-  0.14% )
       227,317,086      branches:u                # 400594036.690 M/sec               ( +-  0.16% )
         4,627,311      branch-misses:u           #    2.04% of all branches          ( +-  0.31% )

            0.8797 +- 0.0203 seconds time elapsed  ( +-  2.31% )
```

There doesn't really seem to be any difference when I did this test. Is there some better way to do one?

**Comment by magik6k** [comment]
Most of this will be I/O. I'd use `testing.B`, load test data like in https://github.com/ipfs/go-ipld-git/blob/master/git_test.go#L22, and then call RawData in a loop `b.N` times

**Comment by sameer** [comment]
Are there any other changes I should make for this PR?

**Pull request merged**